skimage/measure/sopoly.py

```diff
@@ -0,0 +1,1776 @@
+
+    # Fourth column contains the symbol of the cut, and the sixth column the
+    # index of that cut on the corresponding ray.
+    vert_info[all_ord_per_ray, -2] = all_sym_per_ray
+    vert_info[all_ord_per_ray, -1] = all_idx_per_ray
+
+    return vert_info
+
+
+def _merge_new_vertices(vertices, cut_coords, valid_edges, t_coefs):
+    """Merge the new vertices computed from self-intersections and
+    intersections of the polygon with any ray.
+    The newly inserted vertices are sorted according to the coefficient used
+    to compute them.
+
+    Parameters
+    ----------
+    vertices : numpy.ndarray
+        An array containing the coordinates of the polygon vertices.
+    cut_coords : numpy.ndarray
+        A two-dimensional array with coordinates of the intersection vertices.
+    valid_edges : numpy.ndarray
+        A two-dimensional array with the information of the edges that were
+        cut by a ray.
+    t_coefs : numpy.ndarray
+        The coefficients of the parametric rays that define the position of the
+        intersection vertex on its respective edge.
+
+    Returns
+    -------
+    vert_info : numpy.ndarray
+        The set of vetices coordinates with the updated information about the
+        position of intersection vertices.
+    """
+    vert_info = []
+    last_j = 0
+
+    for i in np.unique(valid_edges[:, 0]):
+        vert_info.append(
+            np.hstack((vertices[last_j:i + 1, :],
+                       -np.ones((i - last_j + 1, 1)))))
+        sel_i = np.nonzero(valid_edges[:, 0] == i)[0]
+        sel_r = valid_edges[sel_i, 2]
+
+        ord_idx = np.argsort(t_coefs[sel_i])
+
+        sel_i = sel_i[ord_idx]
+        sel_r = sel_r[ord_idx]
+
+        vert_info.append(np.hstack((cut_coords[sel_i], sel_r.reshape(-1, 1))))
+        last_j = i + 1
+
+    n_vertices = vertices.shape[0]
+    vert_info.append(
+        np.hstack((vertices[last_j:, :], -np.ones((n_vertices - last_j, 1)))))
+    vert_info = np.vstack(vert_info)
+
+    return vert_info
+
+
+def divide_selfoverlapping(coords):
+    """Divide a self-overlapping polygon into non self-overlapping polygons.
+    This implements the algorithm proposed in [1].
+
+    Parameters
+    ----------
+    coords : (N, 2) array
+        Coordinates of the polygon vertices.
+
+    Returns
+    -------
+    sub_polys : list
+        A list of ndarrays with the coordinates of the non
+        self-overlapping polygons obtained from dividing the original polygon.
+
+    References
+    ----------
+    .. [1] Uddipan Mukherjee. (2014). Self-overlapping curves:
+           Analysis and applications. Computer-Aided Design, 46, 227-232.
+           :DOI: https://doi.org/10.1016/j.cad.2013.08.037
+    """
+    # Change the order of the axis to x, y from rr, cc.
+    vertices = coords[:, [1, 0]]
+    max_crest_ids, min_crest_ids, max_crest, min_crest = \
+        _get_crest_ids(vertices, tolerance=2 * np.finfo(np.float32).eps)
```

Comment:
This is duplicate with the body of the `if not is_clockwise` clause. I think it can be moved after that clause:

```python
if not is_clockwise:
    vertices = vertices[::-1, :]

max_crest_ids, min_crest_ids, max_crest, min_crest = _get_crest_ids(
    vertices, tolerance=2 * np.finfo(np.float32).eps
)
```